servers/src/main/java/tachyon/master/file/FileSystemMaster.java

```diff
@@ -1476,6 +1520,183 @@ public void setState(long fileId, SetStateOptions options)
+    Set<Long> scheduledFiles = mWorkerToAsyncPersistFile.get(workerId);
+    for (long fileId : Sets.newHashSet(scheduledFiles)) {
+      InodeFile inode = (InodeFile) mInodeTree.getInodeById(fileId);
+      if (inode.isCompleted()) {
+        toRequestFilePersistence.add(fileId);
+        List<Long> blockIds = Lists.newArrayList();
+        for (FileBlockInfo fileBlockInfo : getFileBlockInfoList(fileId)) {
+          blockIds.add(fileBlockInfo.blockInfo.blockId);
+        }
+
+        PersistFile toCheckpoint = new PersistFile(fileId, blockIds);
+        files.add(toCheckpoint);
+        // update the inode file persisence state
+        inode.setPersistenceState(FilePersistenceState.PERSISTING);
+        scheduledFiles.remove(fileId);
+      }
+    }
+
+    requestFilePersistence(toRequestFilePersistence);
+    // write to journal
+    PersistFilesRequestEntry persistFilesRequest =
+        PersistFilesRequestEntry.newBuilder().addAllFileIds(toRequestFilePersistence).build();
+    writeJournalEntry(
+        JournalEntry.newBuilder().setPersistFilesRequest(persistFilesRequest).build());
+    flushJournal();
+    return files;
+  }
+
+  /**
+   * Request a list of files as being persisted.
+   *
+   * @param fileIds the id of the files
+   * @throws FileDoesNotExistException when a file does not exist
+   */
+  private void requestFilePersistence(List<Long> fileIds) throws FileDoesNotExistException {
+    if (!fileIds.isEmpty()) {
+      LOG.info("Request file persistency: {}", fileIds);
+    }
+    for (long fileId : fileIds) {
+      InodeFile inode = (InodeFile) mInodeTree.getInodeById(fileId);
+      inode.setPersistenceState(FilePersistenceState.PERSISTING);
+    }
+  }
+
+  private void requestFilePersistenceEntry(PersistFilesRequestEntry entry)
+      throws FileDoesNotExistException {
+    requestFilePersistence(entry.getFileIdsList());
+  }
+
+  /**
+   * Instructs a worker to persist the files.
+   *
+   * @param workerId the id of the worker that heartbeats
+   * @return the command for persisting the blocks of a file
+   * @throws FileDoesNotExistException if the file does not exist
+   * @throws InvalidPathException if the file path is invalid
+   */
+  public synchronized PersistCommand workerHeartbeat(long workerId, List<Long> persistedFiles)
+      throws FileDoesNotExistException, InvalidPathException {
+    if (!persistedFiles.isEmpty()) {
+      // notify checkpoint manager the persisted files
+      persistFiles(workerId, persistedFiles);
+    }
+
+    // get the files for the given worker to checkpoint
+    List<PersistFile> filesToCheckpoint = null;
+    filesToCheckpoint = pollToCheckpoint(workerId);
+    if (!filesToCheckpoint.isEmpty()) {
+      LOG.info("Sent files {} to worker {} to persist", filesToCheckpoint, workerId);
+    }
+    return new PersistCommand(CommandType.Persist, filesToCheckpoint);
+  }
+
+  /**
+   * Commits the given list of files as persisted in under file system on a worker.
```

Comment:
It seems like there are two ways to set this persistence state? This method and setState()? This feels like it could get confusing.